rectools/metrics/debias.py

```diff
@@ -0,0 +1,118 @@
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""Debias module."""
+
+import attr
+import pandas as pd
+
+from rectools import Columns
+
+
+@attr.s
+class DebiasConfig:
+    """
+    Config with debias method parameters.
+
+    Parameters
+    ----------
+    iqr_coef : float, default 1.5
+        Coefficient for defining as the maximum value inside the border.
+    random_state : float, default 32
+        Pseudorandom number generator state to control the down-sampling.
+    """
+
+    iqr_coef: float = attr.ib(default=1.5)
+    random_state: int = attr.ib(default=32)
+
+    @property
+    def keyname(self) -> str:
+        """
+        Compose a config's key name based on the iqr_coef and random_state parameters.
+
+        Returns
+        -------
+        str
+            String representing the key name in the format `"{iqr_coef}_{random_state}"`.
+        """
+        return f"{self.iqr_coef}_{self.random_state}"
+
+
+def make_downsample(interactions: pd.DataFrame, debias_config: DebiasConfig) -> pd.DataFrame:
+    """
+    Downsample the size of interactions, excluding some interactions with popular items.
+
+    Algorithm:
+
+        1. Calculate item popularity distribution from interactions;
+        2. Find first (Q1) and third (Q3) quartiles in items popularity distribution;
+        3. Calculate IQR = Q3 - Q1;
+        4. Calculate maximum value inside by formula: Q3 + iqr_coef * IQR;
+        5. Down-sample for all exceeding items in interactions,
+        randomly keeping the maximum group of users to a size not exceeding
+        maximum value inside
+
+    Parameters
+    ----------
+    interactions : pd.DataFrame
+        Table with previous user-item interactions,
+        with columns `Columns.User`, `Columns.Item`.
+    iqr_coef : float, default 1.5
+        Coefficient for defining as the maximum value inside the border.
+    random_state: float, default 32
+        Pseudorandom number generator state to control the down-sampling.
+
+    Returns
+    -------
+    pd.DataFrame
+        Downsampling interactions.
+    """
+    if len(interactions) == 0:
+        return interactions
+
+    item_popularity = interactions[Columns.Item].value_counts()
+
+    quantiles = item_popularity.quantile(q=[0.25, 0.75])
+    q1, q3 = quantiles.loc[0.25], quantiles.loc[0.75]
+    iqr = q3 - q1
+    max_border = int(q3 + debias_config.iqr_coef * iqr)
+
+    item_outside_max_border = item_popularity[item_popularity > max_border].index
+
+    interactions_result = interactions[~interactions[Columns.Item].isin(item_outside_max_border)]
+    interactions_downsampling = interactions[interactions[Columns.Item].isin(item_outside_max_border)]
+
+    interactions_downsampling = (
```

Comment:
I think we can speed this up (please check my code is doing what you want it to do)
<img width="741" alt="image" src="https://github.com/MobileTeleSystems/RecTools/assets/14802567/c94a72ab-6825-4660-8ddc-348670fa22dc">

Reply:
**Done**
But there is more randomness in applying the debias method, so there may be problems with tests.

Reply:
sorry, it's my mistake.
Because I forgot that I made the `random_state` parameter in `DebiasConfig` as `None`, so the results of the `make_debias` function changed in the tests.

Reply:
So, now it's fine? :)

Reply:
Yes)